go/rpcclientlib/obscuro_client.go

```diff
@@ -19,39 +29,228 @@ const (
 	RPCAddViewingKey        = "obscuro_addViewingKey"
 	RPCStopHost             = "obscuro_stopHost"
 	RPCCall                 = "eth_call"
+	RPCGetBalance           = "eth_getBalance"
+	RPCGetTxCount           = "eth_getTransactionCount"
 	RPCGetTxReceipt         = "eth_getTransactionReceipt"
 	RPCSendRawTransaction   = "eth_sendRawTransaction"
 	RPCGetTransactionByHash = "eth_getTransactionByHash"
+
+	// todo: this is a convenience for testnet testing and will eventually be retrieved from the L1
+	enclavePublicKeyHex = "034d3b7e63a8bcd532ee3d1d6ecad9d67fca7821981a044551f0f0cbec74d0bc5e"
+
+	// ViewingKeySignedMsgPrefix is the prefix added when signing the viewing key in MetaMask using the personal_sign
+	// API. Why is this needed? MetaMask has a security feature whereby if you ask it to sign something that looks like
+	// a transaction using the personal_sign API, it modifies the data being signed. The goal is to prevent hackers
+	// from asking a visitor to their website to personal_sign something that is actually a malicious transaction (e.g.
+	// theft of funds). By adding a prefix, the viewing key bytes no longer looks like a transaction hash, and thus get
+	// signed as-is.
+	ViewingKeySignedMsgPrefix = "vk"
 )
 
+// for these methods, the RPC method's requests and responses should be encrypted
+var sensitiveMethods = []string{RPCCall, RPCGetBalance, RPCGetTxCount, RPCGetTxReceipt, RPCSendRawTransaction, RPCGetTransactionByHash}
+
 // Client is used by client applications to interact with the Obscuro node.
 type Client interface {
 	// Call executes the named method via RPC.
 	Call(result interface{}, method string, args ...interface{}) error
 	// Stop closes the client.
 	Stop()
+
+	// SetViewingKey sets the current viewing key on the client to be used for all sensitive request decryption
+	SetViewingKey(viewingKey *ecies.PrivateKey, pubKeyBytes []byte)
+
+	// RegisterViewingKey takes a signature for the public key, it verifies the signed public key matches the currently set private viewing key
+	//	and then submits it to the enclave
+	RegisterViewingKey(signature []byte) error
 }
 
-// A Client implementation that wraps rpc.Client to make calls.
-type clientImpl struct {
```

Comment:
@BedrockSquirrel Not sure why we renamed this?

Reply:
I think impl is a legacy naming convention that comes from languages that required interfaces for everything and it's generally better practice to give implementations a meaningful name where poss.

Client has two implementations, InMem and RPC seems better than InMem and Impl

(But also it was because I stopped using the interface in the wallet extension which you quite rightly called me out on haha)

Reply:
Maybe make it `InMem` and `Network`?